sql-composer/src/types.rs

```diff
@@ -369,6 +410,199 @@ impl fmt::Display for SqlComposition {
+///
+/// Automatically provides a `TryFrom` implementation.
+///
+/// Reads the file at path and parses with `SqlComposition::parse()`.
+/// Equivalent to `SqlComposition::from_path(path)`.
+/// Relative paths are resolved from the directory where
+/// the code is executed.
+///
+/// Can fail reading the contents of path or parsing.
+///
+/// # Examples
+///
+/// TryInto:
+/// ```
+/// use std::convert::TryInto;
+/// use std::path::Path;
+/// use sql_composer::{types::ParsedItemSql,
+///                    error::Result};
+/// fn main() -> Result<()> {
+///   let path = Path::new("src/tests/simple-template.tql");
+///   let stmt: ParsedItemSql = path.try_into()?;
+///   Ok(())
+/// }
+/// ```
+///
+///
+/// TryFrom:
+/// ```
+/// use std::convert::TryFrom;
+/// use std::path::Path;
+/// use sql_composer::{types::ParsedItemSql,
+///                    error::Result};
+/// fn main() -> Result<()> {
+///   let path = Path::new("src/tests/simple-template.tql");
+///   let stmt = ParsedItemSql::try_from(path)?;
+///   Ok(())
+/// }
+/// ```
+///
+/// Concrete `TryFrom` implementations are required because of a
+/// collision with the blanket `TryFrom` implementation.
+/// See Bug 50133 : <https://github.com/rust-lang/rust/issues/50133>
+///
+/// ``` ignore
+/// // Not allowed!
+/// impl ParsedItem<SqlComposition> {
+///     pub fn try_from<P>(path: P) -> Result<Self>
+///         where P: Into<PathBuf> + Debug {
+///             let path = path.into();
+///             let s = fs::read_to_string(&path)?;
+///             SqlComposition::parse(&s, Some(SqlCompositionAlias::from_path(path)))
+///         }
+/// }
+/// ```
+impl TryFrom<&Path> for ParsedItemSql {
+    type Error = Error;
+    fn try_from(path: &Path) -> Result<Self> {
+        SqlComposition::from_path(path)
+    }
+}
+
+/// Implements `TryInto` Trait to convert a PathBuf into a ParsedItemSql.
+///
+/// See TryInfo<Path> implementation for details and motivation.
+///
+/// # Examples
+///
+/// TryInto:
+/// ```
+/// use std::convert::TryInto;
+/// use std::path::PathBuf;
+/// use sql_composer::{types::ParsedItemSql,
+///                    error::Result};
+/// fn main() -> Result<()> {
+///   let path = PathBuf::from("src/tests/simple-template.tql");
+///   let stmt: ParsedItemSql = path.try_into()?;
+///   Ok(())
+/// }
+/// ```
+///
+///
+/// TryFrom:
+/// ```
+/// use std::convert::TryFrom;
+/// use std::path::PathBuf;
+/// use sql_composer::{types::ParsedItemSql,
+///                    error::Result};
+/// fn main() -> Result<()> {
+///   let path = PathBuf::from("src/tests/simple-template.tql");
+///   let stmt = ParsedItemSql::try_from(path)?;
+///   Ok(())
+/// }
+/// ```
+impl TryFrom<PathBuf> for ParsedItemSql {
+    type Error = Error;
+    fn try_from(path: PathBuf) -> Result<Self> {
+        SqlComposition::from_path(path)
+    }
+}
+
+/// Implements `TryInto` Trait to convert a str into a ParsedItemSql.
+///
+/// See TryInfo<Path> implementation for details and motivation.
```

Comment:
1. doc typo: `TryInfo` -> `TryInto`.
2. it should actually be `TryFrom` rather than `TryInto`
3. Would be nice to link to the trait directly.
4. I have no idea what example I wanted to link to since Path doesn't impl TryFrom
5. Much of this is duplicative when viewed in rustdoc. => type in impl gets a link automatically, type signature already has the a->b idea.  rustdoc also seems to pull the doc for the required impl function from the trait, so `try_from` has a stub documentation.
6. trait impl docs are hidden/unexpanded when viewed in rustdoc.  

So let's move this duplicated doc into a module level comment on why we have concrete impls.  And just leave the example doc on each.